cpp/src/arrow/compute/kernels/vector_array_sort.cc

```diff
@@ -173,6 +173,116 @@ class ArrayCompareSorter {
   }
 };
 
+template <>
+class ArrayCompareSorter<DictionaryType> {
+ public:
+  Result<NullPartitionResult> operator()(uint64_t* indices_begin, uint64_t* indices_end,
+                                         const Array& array, int64_t offset,
+                                         const ArraySortOptions& options,
+                                         ExecContext* ctx) {
+    const auto& dict_array = checked_cast<const DictionaryArray&>(array);
+    // TODO: These methods should probably return a const&? They seem capable.
```

Comment:
Seems reasonable. Do you want to open a separate issue for this follow up and tag that issue here?  E.g. `TODO (GH-123): These methods...`

Reply:
https://github.com/apache/arrow/issues/35437 (also included in the TODO)